### PR TITLE
Add support for update entities

### PR DESCRIPTION
**Pull request by ludeeus**
## Proposed change

For https://github.com/home-assistant/core/pull/68248
Preview https://62399243fc89eb000a079abd--home-assistant-gallery.netlify.app/#more-info/update

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/common/entity/compute_state_display.ts [hunk 1]
```diff
@@ -130,6 +131,17 @@ export const computeStateDisplay = (
     }
   }
 
+  if (domain === "update") {
+    // When updating, show "Updating"
+    // When update available, show the version
+    // When update is not available, show "Up-to-date"
+    return compareState === "on"
+      ? isUpdating(stateObj)
+        ? localize("ui.card.update.updating")
```
Comment (ludeeus) [inline src/common/entity/compute_state_display.ts:140]: Should we here use the `in_progress` attribute to show `XX%` instead of "Updating" when the entity supports that?

Comment (zsarnett) [reply src/common/entity/compute_state_display.ts:140]: I think the attribute % would be cool

File: src/data/update.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import {
```
Comment (zsarnett) [inline src/data/update.ts:1]:
```suggestion
import type {
```

**Review by zsarnett — DISMISSED** [review]

File: src/common/entity/compute_state_display.ts (hunk 1 above)
Comment (zsarnett) [reply src/common/entity/compute_state_display.ts:140]: oh that would be nice. Does the state machine update based on the % attribute?

Comment (ludeeus) [reply src/common/entity/compute_state_display.ts:140]: yes

Comment (frenck) [reply src/common/entity/compute_state_display.ts:140]: > Does the state machine update based on the % attribute?

yes/no. If a number is provided for `in_progress` id does, but `in_progress` can also be a boolean value. In case `true` it is updating, but hasn't got specific progress data, `false` means its not updating.

Comment (ludeeus) [reply src/common/entity/compute_state_display.ts:140]: well, in case of not supporting %, it will update when set to true, and then again when set back to false

Comment (ludeeus) [reply src/common/entity/compute_state_display.ts]: implemented in https://github.com/home-assistant/frontend/pull/12059/commits/9b6cfc742ed841a378c108320fa52c68fe4e8732
preview at the bottom here https://623241cfb5bc6a0008a8fbc1--home-assistant-gallery.netlify.app/#more-info/update

Comment (frenck) [reply src/common/entity/compute_state_display.ts]: Looks really cool! I think it would be visually nice if we show a progress indicator or something while updating. E.g., somewhere separate or on the install button.
Not sure if that all is possible, but I think it would visually help if something is moving/animated in the screen to indicate it is busy doing the update

Comment (ludeeus) [reply src/common/entity/compute_state_display.ts]: Everything is possible

File: src/translations/en.json [hunk 3]
```diff
@@ -223,6 +223,11 @@
       "service": {
         "run": "Run"
       },
+      "update": {
+        "updating": "Updating",
```
Comment (ludeeus) [inline src/translations/en.json:227]: Should this be "Installing" instead?

Comment (frenck) [reply src/translations/en.json:227]: Yeah maybe... I called it installing, based on the change/discussions we made some time ago with ESPHome. Downloading, uploading firmware, installing, updating, flashing.

As we are dealing with software, firmware and whatever else, I figured it would be the same context. Thus the component uses "install".

Comment (ludeeus) [reply src/translations/en.json]: Changed to Installing in https://github.com/home-assistant/frontend/commit/79bf80360541c86f6eacecbed5edd3a830d9f739

File: src/data/update.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  release_url: string | null;
+  skipped_version: string | null;
+  title: string | null;
+}
+
+export interface UpdateEntity extends HassEntityBase {
+  attributes: UpdateEntityAttributes;
+}
+
+export const usesProgress = (entity: UpdateEntity): boolean =>
```
Comment (balloob) [inline src/data/update.ts:26]: Since all functions are global in auto-complete, can we name this `updateUsesProgress`

File: src/data/update.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  supportsFeature(entity, UPDATE_SUPPORT_PROGRESS) &&
+  typeof entity.attributes.in_progress === "number";
```
Comment (balloob) [inline src/data/update.ts:28]: Why type of check here ?

File: src/data/update.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+export const canInstallUpdate = (entity: UpdateEntity): boolean =>
+  supportsFeature(entity, UPDATE_SUPPORT_INSTALL) &&
+  entity.attributes.latest_version !== entity.attributes.current_version &&
+  entity.attributes.latest_version !== entity.attributes.skipped_version;
+
+export const updateIsInstalling = (entity: UpdateEntity): boolean =>
+  usesProgress(entity) ? true : (entity.attributes.in_progress as boolean);
```
Comment (balloob) [inline src/data/update.ts:36]: Why having to type cast it to Boolean? If it can be undefined and that should be interpreted as a Boolean, let’s map it as such: `!!in_progress`

Comment (balloob) [reply src/data/update.ts:36]: Can also simplify the statement to be `usesProgress(entity) || !!in_progress`

File: src/dialogs/more-info/controls/more-info-update.ts [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+      ${this.stateObj.attributes.in_progress
+        ? supportsFeature(this.stateObj, UPDATE_SUPPORT_PROGRESS) &&
+          typeof this.stateObj.attributes.in_progress === "number"
+          ? html`<mwc-linear-progress
+              .progress=${this.stateObj.attributes.in_progress / 100}
+              buffer=""
+            ></mwc-linear-progress>`
+          : html`<mwc-linear-progress indeterminate></mwc-linear-progress>`
+        : ""}
+      <h3>${this.stateObj.attributes.title}</h3>
```
Comment (balloob) [inline src/dialogs/more-info/controls/more-info-update.ts:49]: If there is no title, will this render still the margin of h3 ?

File: src/dialogs/more-info/controls/more-info-update.ts [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        </div>
+        <div class="value">${this.stateObj.attributes.current_version}</div>
+      </div>
+      <div class="row">
+        <div class="key">
+          ${this.hass.localize(
+            "ui.dialogs.more_info_control.update.latest_version"
+          )}
+        </div>
+        <div class="value">${this.stateObj.attributes.latest_version}</div>
```
Comment (balloob) [inline src/dialogs/more-info/controls/more-info-update.ts:64]: Should we render a placeholder if no latest version available ? Same with current version that can be null according to type.

File: src/dialogs/more-info/controls/more-info-update.ts [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+      return checkbox.checked;
+    }
+    return true;
+  }
+
+  private _handleInstall(): void {
+    this.hass.callService("update", "install", {
+      entity_id: this.stateObj!.entity_id,
+      backup: this._shouldCreateBackup,
+      version: supportsFeature(this.stateObj!, UPDATE_SUPPORT_SPECIFIC_VERSION)
```
Comment (balloob) [inline src/dialogs/more-info/controls/more-info-update.ts:144]: If an update entity supports specific version, can we not pass `null` to pick last version automatically?

**Review by balloob — DISMISSED** [review]
Nice!

File: src/dialogs/more-info/controls/more-info-update.ts (hunk 8 above)
Comment (ludeeus) [reply src/dialogs/more-info/controls/more-info-update.ts:64]: Sure!, have a suggestion on what that should be?
`localise('states.state.default.unknown')`?

File: src/dialogs/more-info/controls/more-info-update.ts (hunk 7 above)
Comment (ludeeus) [reply src/dialogs/more-info/controls/more-info-update.ts:49]: It will render the h3 element, but it will not take any place
Second example shows that <https://623851bf4fdc510008f01998--home-assistant-gallery.netlify.app/#more-info/update>
But I can add this conditionally if you want

File: src/data/update.ts (hunk 6 above)
Comment (ludeeus) [reply src/data/update.ts:36]: in_progress will always be set, to boolean or number

File: src/dialogs/more-info/controls/more-info-update.ts (hunk 9 above)
Comment (ludeeus) [reply src/dialogs/more-info/controls/more-info-update.ts:144]: I don't see why we can't, I'll test that out and adjust.

File: src/data/update.ts (hunk 5 above)
Comment (ludeeus) [reply src/data/update.ts:28]: To guard against edge-case where the entity set that it supports progress, but are not actually using it.

File: src/dialogs/more-info/controls/more-info-update.ts (hunk 9 above)
Comment (frenck) [reply src/dialogs/more-info/controls/more-info-update.ts:144]: Yeah, we don't have to pass it.
